include/gridtools/reduction/gpu.hpp

```diff
@@ -0,0 +1,318 @@
+                    const unsigned ballot_result = __ballot_sync(0xFFFFFFFF, threadIdx.x < ShmemSize);
+                    if (threadIdx.x < ShmemSize)
+                        res = warp_reduce(f, ballot_result, buff[threadIdx.x]);
+                    if (threadIdx.x == 0)
+                        *(out + blockIdx.x) = res;
+                }
+            };
+
+            template <size_t BlockSize>
+            struct small_warp_kernel {
+                static_assert(BlockSize > 1, GT_INTERNAL_ERROR);
+                static_assert(BlockSize <= 32, GT_INTERNAL_ERROR);
+                template <class F, class T>
+                GT_FUNCTION_DEVICE void operator()(F f, T const *__restrict__ in, T *__restrict__ out) const {
+                    in += blockIdx.x * BlockSize * 2 + threadIdx.x;
+                    T res = warp_reduce(f, 0xFFFFFFFF >> (32 - BlockSize & 31), f(*in, *(in + BlockSize)));
+                    if (threadIdx.x == 0)
+                        *(out + blockIdx.x) = res;
+                }
+            };
+
+            template <class T>
+            using is_shufflable =
+                meta::st_contains<meta::list<int, unsigned, long, unsigned long, long long, unsigned long long, double>,
+                    T>;
+
+            template <size_t BlockSize, class T>
+            using choose_kernel = typename meta::if_c<(BlockSize == 1),
+                // no need to use nor shared memory nighter intrinsics
+                meta::lazy::id<tiny_kernel>,
+                meta::if_<is_shufflable<T>,
+                    meta::if_c<(BlockSize > 32),
+                        // use shared memory and two rounds of warp level intrinsics
+                        meta::lazy::id<big_warp_kernel<BlockSize>>,
+                        // use a single round of warp level intrinsics, no shared memory needed
+                        meta::lazy::id<small_warp_kernel<BlockSize>>>,
+                    // use shared memory only
+                    meta::lazy::id<shmem_kernel<BlockSize>>>>::type;
+
+            template <class Kernel, class F, class T>
+            __global__ void launch(Kernel kernel, F f, T const *__restrict__ in, T *__restrict__ out) {
+                kernel(f, in, out);
+            }
+
+            inline constexpr bool is_pow2(size_t x) { return (x & x - 1) == 0; }
+
+            inline constexpr size_t next_pow2(size_t x) {
+                --x;
+                x |= x >> 1;
+                x |= x >> 2;
+                x |= x >> 4;
+                x |= x >> 8;
+                x |= x >> 16;
+                x |= x >> 32;
+                return ++x;
+            }
+
+            inline constexpr size_t int_log2(size_t val) {
+                assert(val);
+                size_t ret = 0;
+                for (; val > 1; val /= 2)
+                    ret++;
+                return ret;
+            }
+
+            using cpu_final_threshold_t = integral_constant<size_t, 8>;
```

Comment:
Have you benchmarked this? My wild guess would have been that it is at least one GPU cache line or so, but no idea…

Reply:
Not yet benchmarked it seriously. I am adding TODO